### PR TITLE
Add a compilation section to disable address maps

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds a new `Config::generate_address_map` compilation
setting which is used to disable emission of the `.wasmtime.addrmap`
section of compiled artifacts. This section is currently around the size
of the entire `.text` section itself unfortunately and for size reasons
may wish to be omitted. Functionality-wise all that is lost is knowing
the precise wasm module offset address of a faulting instruction or in a
backtrace of instructions. This also means that if the module has DWARF
debugging information available with it Wasmtime isn't able to produce a
filename and line number in the backtrace.

This option remains enabled by default. This option may not be needed in
the future with #3547 perhaps, but in the meantime it seems reasonable
enough to support a configuration mode where the section is entirely
omitted if the smallest module possible is desired.

### CODE REVIEW

File: tests/all/traps.rs [hunk 1]
```diff
@@ -637,3 +637,37 @@ fn multithreaded_traps() -> Result<()> {
 
     Ok(())
 }
+
+#[test]
+#[cfg_attr(all(target_os = "macos", target_arch = "aarch64"), ignore)] // TODO #2808 system libunwind is broken on aarch64
+fn traps_without_addres_map() -> Result<()> {
```
Comment (fitzgen) [inline tests/all/traps.rs:643]:
```suggestion
fn traps_without_address_map() -> Result<()> {
```

**Review by fitzgen — APPROVED** [review]

**Review by pchickey — COMMENTED** [review]
Could we put a guard around `addrs.push(range.clone(), &func.address_map.instructions);`  so we don't even allocate the memory for the address map when we aren't going to omit it?

**Comment by pchickey** [comment]
Also, should this imply `WasmBacktraceDetails::Disable`? What backtrace details will actually be emitted if the addr map isnt present?

**Comment by alexcrichton** [comment]
> Could we put a guard around ...

Ah yes indeed!

> Also, should this imply `WasmBacktraceDetails::Disable`?

While we won't get precise locations within functions (which also means no filenames/line numbers) we do still get the function names in the backtrace due to the `.eh_frame` section. I think because of that we probably don't want to forcibly disable backtraces for now, although in the future we could probably drop the `.eh_frame` section as well (at least on Unix) if it's a size concern.

**Review by pchickey — APPROVED** [review]
Thanks!

**Pull request merged**